### PR TITLE
added config option server-unix-socket-mode

### PR DESCRIPTION
**Pull request by Dansvidania**
started working on solving #1408 

Currently I apply a default (changed to 660) whenever the file mode is not specified, or (I think this part is problematic) whenever the mode provided is invalid (not in octal)

I also still need to validate that the octal provided is not above 777

 @steve-chavez is this the preferred way to set the permission, or should we rather go with a string in the more readable format "rwxrwxrwx" ?

### CODE REVIEW

**Comment by steve-chavez** [comment]
> @steve-chavez is this the preferred way to set the permission, or should we rather go with a string in the more readable format "rwxrwxrwx" ?

@Dansvidania The string is more readable but the octal is more standard(what most unix devs would expect). So I think we should go with octal.

660 default is good. Making sure the value is indeed octal would be even better. Like we do [here for other configs](https://github.com/PostgREST/postgrest/blob/master/main/Main.hs#L175-L181)(we crash if they're invalid).

**Comment by Dansvidania** [comment]
I have added the validation but I am not a huge fan of the code... I am looking forward to your feedback on it

File: main/Main.hs [hunk 1]
```diff
@@ -324,11 +332,12 @@ loadDbUriFile conf = extractDbUri mDbUri
...
-createAndBindSocket :: FilePath -> IO Socket
-createAndBindSocket filePath = do
-  deleteSocketFileIfExist filePath
+createAndBindSocket :: FilePath -> FileMode -> IO Socket
+createAndBindSocket socketFilePath socketFileMode = do
+  deleteSocketFileIfExist socketFilePath
   sock <- socket AF_UNIX Stream defaultProtocol
-  bind sock $ SockAddrUnix filePath
+  bind sock $ SockAddrUnix socketFilePath
+  setFileMode socketFilePath socketFileMode
```
Comment (steve-chavez) [inline main/Main.hs]: How about if we just leave it with the default of `755`(also no breaking change) and let the `setFileMode` be optional:

```haskell
when (isJust socketFileMode) $
  setFileMode socketFilePath $ fromJust socketFileMode
```

Then you could call this function like: `createAndBindSocket (unpack socketAddr) (rightToMaybe socketFileMode)`.

This way you could delete the `432` parts and the validation would be similar for all configs options in the `main` func.

Comment (Dansvidania) [reply main/Main.hs]: @steve-chavez thank you for your feedback, the suggestion made a lot of sense.

Comment (Dansvidania) [reply main/Main.hs]: FYI I ended up changing it to forM_ to comply with linters and then again to mapM_ for better readability

Comment (steve-chavez) [reply main/Main.hs]: Oh, that's better. Lately I've been noticing that there are parts of the code that could be simplified in terms of `Traversable`: 
```hs
setFileMode socketFilePath `traverse_` maybeSocketFileMode
```

But in this case I think the `mapM_` is more readable, so it's good to leave it as it is.

I'll give the PR another review.

File: main/Main.hs [hunk 2]
```diff
@@ -171,6 +175,9 @@ main = do
         . setServerName (toS $ "postgrest/" <> prettyVersion) $
         defaultSettings
 
+  when (isLeft socketFileMode) $
+    panic $ (fromJust . leftToMaybe) socketFileMode
```
Comment (steve-chavez) [inline main/Main.hs]: You could use [whenLeft](http://hackage.haskell.org/package/either-5.0.1.1/docs/Data-Either-Combinators.html#v:whenLeft) here(`whenLeft socketFileMode panic`). You'll have to add the `either` lib to the executable in the cabal file but that's fine.

File: test/SpecHelper.hs [hunk 3]
```diff
@@ -66,6 +66,8 @@ _baseCfg =  -- Connection Settings
   AppConfig mempty "postgrest_test_anonymous" Nothing "test" "localhost" 3000
             -- No user configured Unix Socket
             Nothing
+            -- No user configured Unix Socket file mode (defaults to 660)
+            (Right 432)
```
Comment (steve-chavez) [inline test/SpecHelper.hs]: `493` here by default. Also the comment should be `defaults to 755`·

File: src/PostgREST/Config.hs [hunk 4]
```diff
@@ -248,6 +266,9 @@ readOptions = do
           |## unix socket location
           |## if specified it takes precedence over server-port
           |# server-unix-socket = "/tmp/pgrst.sock"
+          |## unix socket file mode
+          |## if server-unix-socket is used, but an invalid mode or none is provided, 755 is applied
+          |# server-unix-socket-mode = "777"
```
Comment (steve-chavez) [inline src/PostgREST/Config.hs]: Example value could be `server-unix-socket-mode = "755"` to better signal the default.

Also, the comment about `invalid mode` should be removed since now we crash when the value is invalid.

**Comment by steve-chavez** [comment]
See my latest comments, once we clear that I think it's good to merge.

Also, could you help me with updating the docs for this new config option? 

The change would have to be done in https://github.com/PostgREST/postgrest-docs/blob/master/configuration.rst.

**Comment by Dansvidania** [comment]
@steve-chavez thanks for checking again, I should have corrected all the mistakes and changed to `whenLeft` . Is the version condition with a lower boundary only ok for `either` ? i had no idea what would be a reasonable upper boundary

File: postgrest.cabal [hunk 5]
```diff
@@ -97,6 +97,7 @@ executable postgrest
                     , base64-bytestring >= 1 && < 1.1
                     , bytestring        >= 0.10.8 && < 0.11
                     , directory         >= 1.2.6 && < 1.4
+                    , either            >= 5.0.1.1
```
Comment (steve-chavez) [inline postgrest.cabal]: The library uses `either >= 4.4.1 && < 5.1`. You could copy that.

File: main/Main.hs [hunk 6]
```diff
@@ -171,13 +176,16 @@ main = do
         . setServerName (toS $ "postgrest/" <> prettyVersion) $
         defaultSettings
 
+  whenLeft socketFileMode $
```
Comment (steve-chavez) [inline main/Main.hs]:
```suggestion
  whenLeft socketFileMode panic
```

File: main/Main.hs [hunk 7]
```diff
@@ -171,13 +176,16 @@ main = do
...
+    panic $ (fromJust . leftToMaybe) socketFileMode
+
   -- Checks that the provided proxy uri is formated correctly
   when (isMalformedProxyUri $ toS <$> proxy) $
     panic
       "Malformed proxy uri, a correct example: https://example.com:8443/basePath"
 
   -- Checks that the provided jspath is valid
-  when (isLeft roleClaimKey) $
+  whenLeft roleClaimKey $
```
Comment (steve-chavez) [inline main/Main.hs]: I think you changed an unrelated line here.

File: main/Main.hs (hunk 6 above)
Comment (steve-chavez) [reply main/Main.hs]: That would allow you to drop the next line: `panic $ (fromJust . leftToMaybe) socketFileMode`.

File: main/Main.hs (hunk 7 above)
Comment (Dansvidania) [reply main/Main.hs]: I refactored based on the `whenLeft` suggestion you made, to keep it consistent

File: main/Main.hs (hunk 6 above)
Comment (Dansvidania) [reply main/Main.hs]: I was not aware of this usage of panic, thanks.

**Comment by steve-chavez** [comment]
LGTM! :+1: 

Great work done here @Dansvidania. Merging...

**Comment by dwagin** [comment]
The default 755 doesn't make sense, should be 777 or 770.
